### PR TITLE
Fine  grain dynamic shape

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93813
* #94276

A quick prototype to expose problems:
- dynamo.export()'s make_fx is NOT reusing symbols from dynamo. It recreates a new set of symbols. 
- Plumbing dynamic_args via patch("torch._dynamo.config.dynamic_args", ...) is crappy... 
- This API for export() is probably fine, but not sure about torch.compile()
- This doesn't look right `sym_size_2: Sym(16) = torch.ops.aten.sym_size(arg0_1, 2)`. It should directly produce concrete shape. 
- Sym(batch) was produced twice `sym_size` and `sym_size_3`. This should be dedup-ed via ShapeEnv. 

```
class SymIntsAsArg(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x: torch.Tensor, y: torch.Tensor):
        x_new_shape = [x.shape[0] * x.shape[1], x.shape[-1]]
        a =  x.view(x_new_shape)      # SymInt[] as arg

        y_new_shape = [y.shape[0] * y.shape[-1]]
        b =  y.view(y_new_shape)      # SymInt[] as arg

        return a, b

inp = (torch.empty(4, 128, 16), torch.empty(4, 2))
inp_spec = {'x': ('batch', 'seq', None), 'y': ('batch', None)}

opt_model = torch.compile(model, backend="aot_eager", dynamic=True, dynamic_args=model.inp_spec)
```
produces
```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[batch, seq, 16], arg1_1: f32[batch, 2]):
        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_symint.py:31, code: a =  x.view(x_new_shape)      # SymInt[] as arg
        sym_size: Sym(batch) = torch.ops.aten.sym_size(arg0_1, 0)
        sym_size_1: Sym(seq) = torch.ops.aten.sym_size(arg0_1, 1)
        
        # No stacktrace found for following nodes
        mul: Sym(batch*seq) = sym_size * sym_size_1;  sym_size = sym_size_1 = None
        
        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_symint.py:31, code: a =  x.view(x_new_shape)      # SymInt[] as arg
        sym_size_2: Sym(16) = torch.ops.aten.sym_size(arg0_1, 2)
        view: f32[batch*seq, 16] = torch.ops.aten.view.default(arg0_1, [mul, sym_size_2]);  arg0_1 = mul = sym_size_2 = None
        
        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_symint.py:34, code: b =  y.view(y_new_shape)      # SymInt[] as arg
        sym_size_3: Sym(batch) = torch.ops.aten.sym_size(arg1_1, 0)
        sym_size_4: Sym(2) = torch.ops.aten.sym_size(arg1_1, 1)
        
        # No stacktrace found for following nodes
        mul_1: Sym(2*batch) = sym_size_3 * sym_size_4;  sym_size_3 = sym_size_4 = None
        
        # File: /scratch/bahuang/work/repos/pytorch/export_workspace/example_symint.py:34, code: b =  y.view(y_new_shape)      # SymInt[] as arg
        view_1: f32[2*batch] = torch.ops.aten.view.default(arg1_1, [mul_1]);  arg1_1 = mul_1 = None
        return (view, view_1)
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/source.py [hunk 1]
```diff
@@ -63,6 +63,8 @@ def name(self):
 class LocalInputSource(LocalSource):
     pos: int
 
+    dynamic_spec: Optional[Tuple[Optional[str], ...]]   # e.g. ['batch', 'seq', None]
```
Comment (SherlockNoMad) [inline torch/_dynamo/source.py:66]: Not sure if `LocalInputSource` is the right place to store this info.

File: torch/__init__.py [hunk 2]
```diff
@@ -1328,6 +1328,7 @@ def __call__(self, model_, inputs_):
 def compile(model: Optional[Callable] = None, *,
             fullgraph: builtins.bool = False,
             dynamic: builtins.bool = False,
+            dynamic_args = None,
```
Comment (ezyang) [inline torch/__init__.py:1331]: will need dynamic_kwargs too, or maybe a struct that can be passed in here. Type annotation here extremely useful

File: torch/_dynamo/config.py [hunk 3]
```diff
@@ -60,6 +60,9 @@
 # don't specialize on shapes and strides and put shape ops in graph
 dynamic_shapes = os.environ.get("TORCHDYNAMO_DYNAMIC_SHAPES") == "1"
 
+# user provided specs for dynamic args
+dynamic_args = None
```
Comment (ezyang) [inline torch/_dynamo/config.py:64]: This is weird, wasn't expecting the dynamic args to be stashed in global config

File: torch/_dynamo/source.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/source.py:66]: I think it's the wrong place. There should be some data structure in Dynamo which stores the VariableTrackers for local inputs, that data structure seems like the right place.

File: torch/_dynamo/symbolic_convert.py [hunk 4]
```diff
@@ -1656,12 +1656,14 @@ def __init__(
         vars = list(code_options["co_varnames"])
         vars.extend(x for x in self.cell_and_freevars() if x not in vars)
 
+        dynamic_args = torch._dynamo.config.dynamic_args or {}
+
         self.symbolic_locals = collections.OrderedDict(
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:1661]: Yeah, put it on tx (which is self here)

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -761,7 +778,7 @@ def create_symbol(self, val: int, source: Source) -> "sympy.Expr":
 
         # Create a duck sized int if necessary
         if val not in self.val_to_var:
-            sympy_expr = Symbol(f"s{len(self.var_to_val)}", positive=True, integer=True)
+            sympy_expr = Symbol(name or f"s{len(self.var_to_val)}", positive=True, integer=True)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:781]: can't do this because inductor relies on the name being prefixed with s (sorry)

File: torch/_dynamo/source.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/source.py:66]: I see that Source is convenient because we already propagate it through the meta converter. But sorry, it's still the wrong place to put it. Source is a data structure solely for describing where tensors came from, shouldn't hold other stuff.

**Comment by ezyang** [comment]
> This doesn't look right sym_size_2: Sym(16) = torch.ops.aten.sym_size(arg0_1, 2). It should directly produce concrete shape.

I think this is fine for now. In the long term we probably need to apply some orthogonal improvement where we avoid tracing sym_size when its known to be constant.

File: torch/_dynamo/source.py (hunk 1 above)
Comment (SherlockNoMad) [reply torch/_dynamo/source.py]: If we don't reuse LocalInputSource, then I have to plumb it through 7 files https://github.com/pytorch/pytorch/pull/93820
which is uhhhh... 

Also, only input tensor could have dynamic_spec, adding dynamic_spec to VariableBuilder and VariableTracker doesn't looks correct. 

Maybe we should extend the meaning of LocalInputSource?

File: torch/__init__.py (hunk 2 above)
Comment (jansel) [reply torch/__init__.py:1331]: I think we really want per-dimension tagging, not just at the arg level.  We could reuse some FX type annotations:
```
from torch.fx.tensor_type import TensorType, Dyn

def foo(x: TensorType[Dyn, 100]):
   ...
```

File: torch/_dynamo/config.py (hunk 3 above)
Comment (jansel) [inline torch/_dynamo/config.py:64]: Don't put this in the config, shape env seems better.

File: torch/__init__.py (hunk 2 above)
Comment (ezyang) [reply torch/__init__.py:1331]: Sherlock's design does support per dim

File: torch/__init__.py [hunk 6]
```diff
@@ -225,6 +227,8 @@ def _load_global_deps():
 if TYPE_CHECKING:
     import torch._C as _C
 
+TensorSpec = List[Optional[str]]
```
Comment (SherlockNoMad) [inline torch/__init__.py:230]: @jansel, @ezyang 

This is not exactly the right storage type, but I also feel torch.fx.TensorType is not exactly right either... 

I think we need something that allow user writing ('batch', Dyn, 8, Static)
'batch': dynamic, and please use 'batch' as the symbol name
Dyn: dynamic, and I don't care about the symbol name
8: static, and I know this must be 8, and compiler please error out if this is not 8
Static: static, but I don't know what's the actual value, please use the shape from sample input.

Is this over engineered?

File: torch/_dynamo/variables/builder.py [hunk 7]
```diff
@@ -161,12 +161,14 @@ def __init__(
         self,
         tx,
         source: Source,
+        dynamic_spec=None,
```
Comment (SherlockNoMad) [inline torch/_dynamo/variables/builder.py:164]: It feels weird that I am adding this field here, for all python Variables, where this only makes sense for TensorVariable. Something's wrong with this abstraction... 

Maybe I should minor the class structure for Source. 

```
@dataclass
class TensorSpec(VariableSpec)
     dynamic_spec: ...
```

File: torch/_dynamo/source.py [hunk 8]
```diff
@@ -270,6 +273,8 @@ def name(self):
             else:
                 return f"{self.base.name()}[{self.index!r}]"
 
+    def __hash__(self):
```
Comment (SherlockNoMad) [inline torch/_dynamo/source.py:276]: Is LocalInputSource and GetItemSource enough to cover the pytree-able input structs?

File: torch/_dynamo/config.py (hunk 3 above)
Comment (SherlockNoMad) [reply torch/_dynamo/config.py]: Moved to ShapeEnv.

File: torch/__init__.py (hunk 6 above)
Comment (jansel) [reply torch/__init__.py:230]: I'd suggest dropping support for custom names.  They seem too complicated and likely to break things down stream.  Custom names are also not composable, you might have two different codebases sharing the same name, and it is not clear if they should be implicitly merged into one or name mangled to make them different.

I'd prefer something more like debug annotations and source information, which would also be useful for shapes coming from non-inputs and a more general solution.  That seems like a separate PR though, and I wouldn't try to bundle it with this change.

Without the custom naming stuff,  I think we could just add `Static` to `torch.fx.TensorType` and get to your proposal.

---

On the UX side, there are a few options:
1) Invent some new data structure like `dynamic_args` that mirrors the args/kwargs/params.  We could pass `torch.fx.TensorType` into this data structure, or create something new like you do here.
2) Push the work of picking dynamic/static dims into a user-defined callback function, that is called once per Tensor by VariableBuilder.
3) Piggyback on Python type annotations like was done for the FX gradual shape typing work.
4) Use tensor subclassing to attach shape annotations directly to the Tensor object.

I dislike option 1 the most. I worry the data structures will be brittle (support for nesting/kwargs/params etc).  And I also want the annotations to be close to the actual data.  I think a side data structure is suboptimal UX.

2/3/4 all seem like reasonable choices to me.

In terms of easy to implement 4 might actually be easiest since you don't need to do any of the plumbing, just a small tweak to VariableBuilder.  It would also add support for non-inputs (and nested inputs) for free.

Comment (ezyang) [reply torch/__init__.py:230]: >  Custom names are also not composable, you might have two different codebases sharing the same name, and it is not clear if they should be implicitly merged into one or name mangled to make them different.

FWIW, the behavior I'd expect is that the names are scoped only to the torch.compile expression they're associated with. So really it's just a way of conveniently specifying equality relationships between dynamic dimensions at the call site. I don't really have any bone one way or another with whether or not we should keep custom names though.

Comment (voznesenskym) [reply torch/__init__.py:230]: Naming feels like a refinement / syntatic sugar once you have the core data structure designed. 

To add another option to @jansel's list above - composition feels like the best choice. Wrap a user tensor in a DynamicTensor() instead of composing, (1) to force the export system to unpack it and (2) to not deal with the annoying subclassing problems we have today for when you want to fakify a tensor subclass. 

`DynamicTensor(tensor, dynamic_dims=[], static_dims=[])` 

Also, having both static and dynamic feels like maybe too much now? Perhaps dynamic only, and assume static is fine, due to the overall specializing nature of export?

File: torch/_dynamo/variables/builder.py (hunk 7 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:164]: This is the wrong place for this - you want this alongside (or just directly on) the `value` on `__call___` below

File: torch/__init__.py (hunk 6 above)
Comment (jansel) [reply torch/__init__.py:230]: Wrapping seems pretty similar to subclassing.  When we convert the subclass to fake tensors we can just drop/ignore the subclass and treat them like normal tensors.  The fake tensor constructor just needs to copy over the dynamic bits into the symsizes.

Comment (voznesenskym) [reply torch/__init__.py:230]: The problem is now FakeTensor has to know about these, and if someone ever adds overridden methods to it, you need to do simulation or lose the expected behavior.

Comment (ezyang) [reply torch/__init__.py:230]: Sorry, isn't DynamicTensor the subclass option 3 in https://docs.google.com/document/d/1yUik4D3LlBMGV1OIh96FEvGr6AcigrMbj8O0MXZPtdY/edit#heading=h.4oz3g6vsy0e8 which we concluded is too difficult to implement correctly (because you're obligated to propagate it through eager mode)

Comment (voznesenskym) [reply torch/__init__.py:230]: > Sorry, isn't DynamicTensor the subclass option 3 in https://docs.google.com/document/d/1yUik4D3LlBMGV1OIh96FEvGr6AcigrMbj8O0MXZPtdY/edit#heading=h.4oz3g6vsy0e8 which we concluded is too difficult to implement correctly (because you're obligated to propagate it through eager mode)

The distinction is that this is (1) Export/whole_graph only (2) a meaningless notion outside the compiler stack, so no eager prop needed.

Comment (voznesenskym) [reply torch/__init__.py:230]: Still need a param story tho @SherlockNoMad

Comment (ezyang) [reply torch/__init__.py:230]: I struggle to imagine a situation when you need to mark parameter dims as dynamic.

In any case, I think it's clear we should call a meeting on this.

Comment (SherlockNoMad) [reply torch/__init__.py:230]: @voznesenskym 

It's unclear if this is only needed for export path. In our previous discussion, we assumed that this could also useful for torch.compile() path, as having static shape would unlock some potential optimization downstream. 

Ideally, we want to have the same UX for torch.export and torch.compile, but I am fine if we end up with an export-only UX to begin with. 

I don't think we need to worry too much about dynamic parameter dims case. For export, parameters must be concrete. Otherwise, there is no way to serialize a parameter. For torch.compile, cannot think of a case where we need to change parameter shape on the fly.

Comment (jansel) [reply torch/__init__.py:230]: First time seeing that doc.  Based on reading it I think we should go with subclasses.   It is the only option I see that has a hope of working with graph breaks where you need to know which inputs to a graph-break-resume function are dynamic.  Dynamic shapes is *not* an export-only thing where whole graphs is an ok requirement.

Implementation wise, I don't think it will be hard.  We don't actually need to preserve the subclasses when we convert to FakeTensor and trace.  When converting DynamicTensor to FakeTensor just set the dynamic shapes on the fake tensor properly and treat it like a Tensor.

Comment (voznesenskym) [reply torch/__init__.py:230]: > It's unclear if this is only needed for export path. 

No, but I think we are taking nopython=True as a firm requirement?

> Dynamic shapes is not an export-only thing where whole graphs is an ok requirement.
> It is the only option I see that has a hope of working with graph breaks

I thought we had decided firmly against propagation across eager? cc: @Chillee

Comment (ezyang) [reply torch/__init__.py:230]: We'll talk about this tomorrow.
